compiler/rustc_abi/src/lib.rs

```diff
@@ -1118,6 +1118,13 @@ rustc_index::newtype_index! {
     pub struct FieldIdx {}
 }
 
+/// `offset_of` can traverse fields and enum variants and should keep track of which is which.
```

Comment:
Hmm, `[Variant(…), Variant(…)]` is impossible, right?  Would it be worth trying to include that in the representation?  Or maybe just having it always be `(VariantIdx, FieldIdx)`, since it can use `FIRST_VARIANT` even for `struct`s?